File: packages/houdini-lsp/src/extract.parity.test.ts
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { expect, test } from 'vitest'

import { extract_blocks } from './extract'

// The Go extractor (houdini-core/plugin/documents/processFile.go) is the authority;
// this module re-implements it for unsaved editor buffers. Both run the shared
// corpus against one golden — TestExtractParityGolden covers the Go side. If either
// half fails, the extractors have drifted and inline-document positions are wrong.
const FIXTURES = fileURLToPath(
	new URL('../../houdini-core/plugin/documents/testdata/extract-parity/', import.meta.url)
)

test('extract_blocks matches the Go extractor on the shared corpus', () => {
	const text = readFileSync(`${FIXTURES}corpus.tsx`, 'utf-8')
	const expected = JSON.parse(readFileSync(`${FIXTURES}expected.json`, 'utf-8'))

	const got = extract_blocks(text, 'file:///corpus.tsx').map((block) => ({
		content: block.content,
		row: block.offsetLine,
		column: block.offsetColumn,
		prop: block.prop ?? '',
	}))

	expect(got).toEqual(expected)
})

File: packages/houdini-core/plugin/documents/extract_parity_test.go
package documents_test

import (
	"encoding/json"
	"os"
	"testing"

	"github.com/spf13/afero"
	"github.com/stretchr/testify/require"

	"code.houdinigraphql.com/packages/houdini-core/plugin/documents"
)

// The language server re-implements this package's extraction in TypeScript so it
// can extract from unsaved editor buffers (packages/houdini-lsp/src/extract.ts).
// Both implementations run the shared corpus and must produce the same golden —
// if this test and the vitest parity suite disagree, the two extractors have
// drifted and inline-document positions in the editor are wrong.
func TestExtractParityGolden(t *testing.T) {
	ch := make(chan documents.DiscoveredDocument, 100)
	err := documents.ProcessFile(afero.NewOsFs(), "testdata/extract-parity/corpus.txt", ch)
	require.Nil(t, err)
	close(ch)

	type doc struct {
		Content string `json:"content"`
		Row     int    `json:"row"`
		Column  int    `json:"column"`
		Prop    string `json:"prop"`
	}
	got := []doc{}
	for d := range ch {
		got = append(got, doc{Content: d.Content, Row: d.OffsetRow, Column: d.OffsetColumn, Prop: d.Prop})
	}

	expected, readErr := os.ReadFile("testdata/extract-parity/expected.json")
	require.NoError(t, readErr)
	gotJSON, marshalErr := json.MarshalIndent(got, "", "    ")
	require.NoError(t, marshalErr)
	require.JSONEq(t, string(expected), string(gotJSON))
}
